src/test/java/uk/gov/pay/ledger/util/pagination/PaginationBuilderTest.java

```diff
@@ -138,4 +141,27 @@ public void shouldNotGenerateNextLink_whenLimitTotalIsSetAndCountIsLessThanPageS
         builder = builder.buildResponse();
         assertThat(builder.getNextLink(), is(nullValue()));
     }
-}
\ No newline at end of file
+    
+    @Test
+    public void shouldGenerateSelfLinkWithOnlyPrescribedSpecialCharactersURLEncoded(){
+        Map<String, Object> valuesForReferenceParameter = Map.of(
+                "{ref", "reference=%7Bref",
+                "[ref", "reference=%5Bref",
+                "r{e{f}{", "reference=r%7Be%7Bf%7D%7B",
+                "ref=", "reference=ref=",
+                "ref&", "reference=ref&",
+                "ref@", "reference=ref%40"
+        );
+        for (Map.Entry<String, Object> entry : valuesForReferenceParameter.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+            transactionSearchParams.setReference(key);
+            PaginationBuilder builder = new PaginationBuilder(transactionSearchParams, mockedUriInfo)
+                    .withTotalCount(120L)
+                    .withCount(9L)
+                    .buildResponse();
+            assertThat(builder.getSelfLink().toString().contains(value.toString()), is(true));
+            assertDoesNotThrow(builder::getSelfLink);
```

Comment:
I think it would be good to also check that the URL ends up being what we expect (since it’s possible for this to fail in a way that gives us a valid but incorrect URL)

Reply:
I was wondering if it's better to use contains() or equals() here? I guess equals() is less likely to deliver false positives?

Reply:
`equals(…)` sounds better. You could do something like only look at the substring after the first `?` to reduce the clutter.